pkg/csi/cinder/controllerserver.go

```diff
@@ -382,15 +477,92 @@ func (cs *controllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateS
 		klog.V(3).Infof("CreateSnapshot %s from volume with ID: %s", name, volumeID)
 	}
 
-	ctime := timestamppb.New(snap.CreatedAt)
-	if err := ctime.CheckValid(); err != nil {
-		klog.Errorf("Error to convert time to timestamp: %v", err)
+	if !skipSnapshot {
+		ctime = timestamppb.New(snap.CreatedAt)
+		if err = ctime.CheckValid(); err != nil {
+			klog.Errorf("Error to convert time to timestamp: %v", err)
+		}
+
+		err = cs.Cloud.WaitSnapshotReady(snap.ID)
+		if err != nil {
+			klog.Errorf("Failed to WaitSnapshotReady: %v", err)
+			return nil, status.Errorf(codes.Internal, "CreateSnapshot failed with error %v", err)
+		}
+
+		snapSize = snap.Size
 	}
+	// If snapshotType is 'backup', create a backup from the snapshot and delete the snapshot.
+	if snapshotType == "backup" {
 
-	err = cs.Cloud.WaitSnapshotReady(snap.ID)
-	if err != nil {
-		klog.Errorf("Failed to WaitSnapshotReady: %v", err)
-		return nil, status.Errorf(codes.Internal, "CreateSnapshot failed with error %v", err)
+		// Add cluster ID to the snapshot metadata
+		properties := map[string]string{cinderCSIClusterIDKey: cs.Driver.cluster}
+
+		// see https://github.com/kubernetes-csi/external-snapshotter/pull/375/
+		// Also, we don't want to tag every param but we still want to send the
+		// 'force-create' flag to openstack layer so that we will honor the
+		// force create functions
+		for _, mKey := range []string{"csi.storage.k8s.io/volumesnapshot/name", "csi.storage.k8s.io/volumesnapshot/namespace", "csi.storage.k8s.io/volumesnapshotcontent/name", openstack.SnapshotForceCreate, openstack.SnapshotType, openstack.BackupMaxDurationPerGB} {
+			if v, ok := req.Parameters[mKey]; ok {
+				properties[mKey] = v
+			}
+		}
+
+		// We repeat the check from the snapshot creation in case Cinder-CSI gets restarted while a backup is in progress.
+		if len(backups) == 1 {
```

Comment:
try to avoid `if/then-else-if/else` blocks.

You can do the following:

```golang
if len(backups) > 1 {
   return err
}

if len(backup) == 0 {
      // create a backup
} else {
    // do smth
}
}
```